src/common/NetworkManager.ts

```diff
@@ -402,7 +402,6 @@ export class NetworkManager extends EventEmitter {
 
     if (requestWillBeSentEvent) {
       this._onRequest(requestWillBeSentEvent, interceptionId);
-      this._requestIdToRequestWillBeSentEvent.delete(requestId);
```

Comment:
I believe this is safe to remove because `_forgetRequest` cleans everything up at the end.

Reply:
LGTM, although I'd feel safer if this was accompanied by one or two unit tests for redirect requests

Reply:
I'm wondering if this plays nicely with redirect requests

Reply:
@Androbin Can you think of additional tests beyond the existing redirect tests? #7764 added a lot of tests for these cases, I think this line may have been inadvertently missed because there was no test specifically for the double pause bug (crbug.com/1196004) which actually extends beyond just stylesheets. @josepharhar do you have thoughts on additional redirect tests and what they might look like?

Reply:
I should also add that @hemanthreddyk discovered that removal of this line causes an `Unexpected extraInfo events for request` in some cases (like browsing to google.com). This check was added in #7764. I have not looked into the cause, but unless we fix it, this PR will trade stalled requests for exceptions thrown :)

Reply:
> #7764 added a lot of tests for these cases, [...]

Those tests aren't even request interception enabled

> [...] specifically for the double pause bug (crbug.com/1196004) which actually extends beyond just stylesheets.

AFAIK that issue is specific to font requests, is it not?

Reply:
I have decided that I'm going to trust redirect requests to be fine, as I trust the guard clause (added in https://github.com/puppeteer/puppeteer/pull/7060/commits/5f183e68e691acc7b8115f7c09345921214e4a72).

Reply:
A non-stylesheet double-pause can be observed in `main` by running this script.

```ts
const puppeteer = require('.');

(async () => {
  const browser = await puppeteer.launch({ headless: false });
  const page = await browser.newPage();
  await page.setRequestInterception(true);
  page.on('request', (interceptedRequest) => {
    interceptedRequest.continue();
  });
  await page.goto('https://www.google.com');
  await browser.close();
})();
```

Output:

```
_onRequestWillBeSent {
  "requestId": "0BA1C765EC75BF5F569B02C7E5390115",
  "loaderId": "0BA1C765EC75BF5F569B02C7E5390115",
  "documentURL": "https://www.google.com/",
  "request": {
    "url": "https://www.google.com/",
    "method": "GET",
    "headers": {
      "sec-ch-ua": "\"Chromium\";v=\"97\", \" Not;A Brand\";v=\"99\"",
      "sec-ch-ua-mobile": "?0",
      "sec-ch-ua-platform": "\"macOS\"",
      "Upgrade-Insecure-Requests": "1",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4691.0 Safari/537.36"
    },
    "mixedContentType": "none",
    "initialPriority": "VeryHigh",
    "referrerPolicy": "strict-origin-when-cross-origin",
    "isSameSite": true
  },
  "timestamp": 320492.464291,
  "wallTime": 1638119448.560428,
  "initiator": {
    "type": "other"
  },
  "redirectHasExtraInfo": false,
  "type": "Document",
  "frameId": "1F5C0D6493B1B4635224E6A6774EF92E",
  "hasUserGesture": false
}
_onRequestPaused {
  "requestId": "interception-job-1.0",
  "request": {
    "url": "https://www.google.com/",
    "method": "GET",
    "headers": {
      "sec-ch-ua": "\"Chromium\";v=\"97\", \" Not;A Brand\";v=\"99\"",
      "sec-ch-ua-mobile": "?0",
      "sec-ch-ua-platform": "\"macOS\"",
      "Upgrade-Insecure-Requests": "1",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4691.0 Safari/537.36",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"
    },
    "initialPriority": "VeryHigh",
    "referrerPolicy": "strict-origin-when-cross-origin"
  },
  "frameId": "1F5C0D6493B1B4635224E6A6774EF92E",
  "resourceType": "Document",
  "networkId": "0BA1C765EC75BF5F569B02C7E5390115"
}
_onRequestPaused {
  "requestId": "interception-job-2.0",
  "request": {
    "url": "https://www.google.com/",
    "method": "GET",
    "headers": {
      "sec-ch-ua": "\"Chromium\";v=\"97\", \" Not;A Brand\";v=\"99\"",
      "sec-ch-ua-mobile": "?0",
      "sec-ch-ua-platform": "\"macOS\"",
      "Upgrade-Insecure-Requests": "1",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4691.0 Safari/537.36",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "sec-ch-ua-full-version": "\"97.0.4691.0\"",
      "sec-ch-ua-arch": "\"arm\"",
      "sec-ch-ua-platform-version": "\"12.0.1\"",
      "sec-ch-ua-model": "\"\""
    },
    "initialPriority": "VeryHigh",
    "referrerPolicy": "strict-origin-when-cross-origin"
  },
  "frameId": "1F5C0D6493B1B4635224E6A6774EF92E",
  "resourceType": "Document",
  "networkId": "0BA1C765EC75BF5F569B02C7E5390115"
}
```

Reply:
Related to @hemanthreddyk's discovery, here is a case where CDP sends an unexpected `Network.responseReceivedExtraInfo` event. Notice that the response event says `"hasExtraInfo": false,` even though one was received:

```ts
const puppeteer = require('.');

(async () => {
  const browser = await puppeteer.launch({
    headless: false,
  });
  const page = await browser.newPage();
  await page.goto('https://www.google.com', {
    waitUntil: 'networkidle0',
  });
  await browser.close();
})();
```

Output:

```
_onRequestWillBeSent {
  "requestId": "3424.21",
  "loaderId": "912BCB9C93F7C622C5615297D7A27A01",
  "documentURL": "https://www.google.com/",
  "request": {
    "url": "https://adservice.google.com/adsid/google/ui",
    "method": "GET",
    "headers": {
      "sec-ch-ua": "\"Chromium\";v=\"97\", \" Not;A Brand\";v=\"99\"",
      "Referer": "https://www.google.com/",
      "sec-ch-ua-mobile": "?0",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4691.0 Safari/537.36",
      "sec-ch-ua-platform": "\"macOS\""
    },
    "mixedContentType": "none",
    "initialPriority": "Low",
    "referrerPolicy": "origin",
    "isSameSite": true
  },
  "timestamp": 320813.483986,
  "wallTime": 1638119769.574235,
  "initiator": {
    "type": "other"
  },
  "redirectHasExtraInfo": false,
  "type": "Image",
  "frameId": "1F0FED4D77A29AD2A366E039FF44F452",
  "hasUserGesture": false
}
_onResponseReceivedExtraInfo {
  "requestId": "3424.21",
  "blockedCookies": [],
  "headers": {
    "p3p": "CP=\"This is not a P3P policy! See http://support.google.com/accounts/answer/151657 for more info.\"",
    "timing-allow-origin": "*",
    "cross-origin-resource-policy": "cross-origin",
    "cache-control": "private, max-age=15",
    "content-type": "text/html; charset=UTF-8",
    "x-content-type-options": "nosniff",
    "date": "Sun, 28 Nov 2021 17:16:10 GMT",
    "server": "cafe",
    "content-length": "0",
    "x-xss-protection": "0",
    "alt-svc": "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
  },
  "resourceIPAddressSpace": "Public",
  "statusCode": 204
}
_onResponseReceived {
  "requestId": "3424.21",
  "loaderId": "912BCB9C93F7C622C5615297D7A27A01",
  "timestamp": 320813.863494,
  "type": "Image",
  "response": {
    "url": "https://adservice.google.com/adsid/google/ui",
    "status": 204,
    "statusText": "",
    "headers": {},
    "mimeType": "text/html",
    "connectionReused": false,
    "connectionId": 188,
    "remoteIPAddress": "142.250.191.162",
    "remotePort": 443,
    "fromDiskCache": false,
    "fromServiceWorker": false,
    "fromPrefetchCache": false,
    "encodedDataLength": 387,
    "timing": {
      "requestTime": 320813.484787,
      "proxyStart": -1,
      "proxyEnd": -1,
      "dnsStart": 0.197,
      "dnsEnd": 98.239,
      "connectStart": 98.239,
      "connectEnd": 277.558,
      "sslStart": 179.533,
      "sslEnd": 277.553,
      "workerStart": -1,
      "workerReady": -1,
      "workerFetchStart": -1,
      "workerRespondWithSettled": -1,
      "sendStart": 277.805,
      "sendEnd": 278.001,
      "pushStart": 0,
      "pushEnd": 0,
      "receiveHeadersEnd": 376.95
    },
    "responseTime": 1638119769951.583,
    "protocol": "h2",
    "securityState": "secure",
    "securityDetails": {
      "protocol": "TLS 1.3",
      "keyExchange": "",
      "keyExchangeGroup": "X25519",
      "cipher": "AES_128_GCM",
      "certificateId": 0,
      "subjectName": "*.google.com",
      "sanList": [...]
      ],
      "issuer": "GTS CA 1C3",
      "validFrom": 1635733192,
      "validTo": 1642990791,
      "signedCertificateTimestampList": [],
      "certificateTransparencyCompliance": "unknown"
    }
  },
  "hasExtraInfo": false,
  "frameId": "1F0FED4D77A29AD2A366E039FF44F452"
}
```

Reply:
Damn

Reply:
so this PR introduces a new test and does not actually change the behavior? or is removing `this._requestIdToRequestWillBeSentEvent.delete(requestId);` critical for the workaround for the double pause issue?

Reply:
@OrKoN The test included with this PR will fail in `main` because the double pause issue will cause the request to hang. Removing `this._requestIdToRequestWillBeSentEvent.delete(requestId);` makes it pass, but it also exposes the unrelated but similar problem of CDP sending a `Network.responseReceivedExtraInfo` before `Network.responseReceived`. We may want to consider fixing both in this PR, otherwise this PR is likely to just shift the bug reports to the extraInfo problem. 

As far as I can tell, CDP doesn't guarantee the order of `Network.responseReceived` `Network.responseReceivedExtraInfo`.  So there are two possibilities:

```
// Scenario 1 - no extraInfo
Network.responseReceived
   hasExtraInfo: false

// Scenario 2 - has extraInfo
Network.responseReceived
   hasExtraInfo: true
Network.responseReceivedExtraInfo

// Scenario 3 - the one we don't cover - extraInfo comes first
Network.responseReceivedExtraInfo
Network.responseReceived
   hasExtraInfo: false <-- note this is false because there is no forthcoming extraInfo event since it has already fired
```